expr/funcs.go

```diff
@@ -0,0 +1,44 @@
+package expr
+
+import "github.com/raintank/metrictank/api/models"
+
+type argType uint8
+
+// argument type. potentially not as strict as reality (e.g. movingAverage windowsize is categorized as a str) that's why we have the extra validation step
+const (
+	series argType = iota
+	seriesList
+	seriesLists // one or multiple seriesLists
+	integer     // number without decimals
+	float       // number potentially with decimals
+	str         // string
+)
+
+type Func interface {
+	Signature() ([]argType, []argType)
+	// what can be assumed to have been pre-validated: len of args, and basic types (e.g. seriesList)
+	Init([]*expr) error                                                  // initialize and validate arguments (expressions given by user), for functions that have specific requirements
+	Depends(from, to uint32) (uint32, uint32)                            // allows a func to express its dependencies
```

Comment:
couldn't find a better way to phrase it, but the idea is that an invocation like `movingAverage(foo, "5min")` with from=x and to=y can declare it "depends" on data from=x-5min to y as input.